Light linking

mattigruener opened the pull request:
Hey John :)

Ok, so this is one of those PRs that will start a discussion about how to actually do this. 

I have a simple, but working prototype for the light linking internals that I would like to get some early feedback on in case that you don't like any of this at all. It will need some cleaning up in some areas, but I guess for the most part this is how I could imagine this being implemented. Can I pick your brain on what you'd like to have changed? I am sure I'm violating some conventions again.

Building a UI for this will be a pain at some point too, but I'm ignoring that for now. I will send you the example scene that I've been using for this via eMail, because it contains some IE nodes (should have stripped them out, but I gotta run).

Thanks for your help :)
Matti.

mattigruener commented:
Hey John,

thanks for the feedback. I'll look into addressing most of it today. Just as an additional question that's also related to the tests: How do you feel about adding a `ScenePlug::evaluateSetExpression()` as an addition to `ScenePlug::set()` and `ScenePlug::setNames()`? I was talking to Dan about it and we both felt that doing it that way could be useful as a python binding? A `ScenePlug` is necessary anyway for the `SetExpression` to make any sense. The test for the second stage (evaluating the expression) could be done very easily that way.

I'll comment on the rest of your feedback as well - just wanted to get this out there while you're still around maybe :)

Matti.

mattigruener commented on src/GafferScene/SetExpression.cpp:
```diff
@@ -0,0 +1,276 @@
+
+result_type operator()(std::string setName) const
+{
+	// TODO: handle case in which the setName given is not a valid set
```
Mh ok. I guess I had a case in mind where a user would add an expression like this:

`(mySetA & mSetB)`  _Note the typo_

What you're proposing is to just silently return an empty PathMatcher, right? Would you be opposed to at least printing a warning that the requested set couldn't be found? What I had in mind would look like this:

```
result_type operator()( std::string setName ) const
{
  const std::vector<IECore::InternedString> &setNames = m_scene->setNames()->readable();
  if( std::find( setNames.begin(), setNames.end(), IECore::InternedString( setName ) ) != setNames.end() )
  {
    return m_scene->set( setName )->readable();
  }
  else
  {
    throw IECore::Exception( boost::str( boost::format( "Set %s is not available in SetExpression." ) % setName ) );
  }
}
```

johnhaddon replied:
Ah, I see. I wonder if this will actually be annoying for users though, particularly when lookdev and light rigs might be worked on separately by different people, or when some lights might be deleted/disabled in some render passes, meaning the set referenced by the lookdev doesn't exist any more?

mattigruener commented on src/GafferScene/SetExpression.cpp:
```diff
@@ -0,0 +1,276 @@
+	}
+	else
+	{
+		printf("No! No! No! You're doing it wrong! I'll give you something equally as wrong!");
```
I guess we wouldn't even get to this point, because the parser wouldn't recognize any other operator and bail on us. That needs a proper error btw - potentially with an indication of how far into the SetExpression parsing succeed?

```
Could not fully parse SetExpression "(mySetA | mySetB) % mySetC"
                                                       ^
```

The current message is definitely nothing more than a placeholder. I hadn't considered that we prbly don't even have to handle that case in the `AstEvaluator`...

johnhaddon replied:
Ah, I see. Maybe the real issue then is that we're using `char` in BinaryOp, so our AST allows greater freedom than our parser does? How about we add an `enum Op { Intersection, Union }` to BinaryOp, and store one of those instead of a `char`? That way we can use a `switch` instead of `if...else...`, and the compiler will warn for any unhandled cases if we add further operations in the future.

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -769,6 +773,29 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+
+				if( m_linkedLights )
+				{
+					const std::vector<std::string> &lightNamesWritable = m_linkedLights->readable();
```
Ah, yeah - I used `->writable()` before and forgot to fix the variable name when I fixed the method call.

mattigruener commented on src/GafferScene/SetExpression.cpp:
```diff
@@ -0,0 +1,276 @@
+
+		for(PathMatcher::Iterator it = left.begin(); it != left.end(); ++it)
+		{
+			if( right.find( *it ) != right.end() )
```
I'm not sure I follow. I had tried the `RawInterator` first as that is what find() returns, but it gave me exactly the problem you describe: it matched all kinds of ancestor locations. When I switched to the current implementation, which should only visit the actual locations, that problem went away.

I just did another test using the added `GafferSceneTest.PathMatcherTest.PathMatcherTest.testIntersection()` unittest and checked both `Interator` and `RawInterator`. Iterator works fine while `RawIterator` gives me this - which I kinda expected:

```
Traceback (most recent call last):
  File "/home/mattig/apps/gaffer/0.31.0.0dev/cent7.x86_64/cortex/9/gaffer/python/GafferSceneTest/PathMatcherTest.py", line 1011, in testIntersection
    self.assertEqual( m3.paths(), [ "/a/b/c/d/myTest" ] )
AssertionError: Lists differ: ['/', '/a', '/a/b', '/a/b/c', ... != ['/a/b/c/d/myTest']

First differing element 0:
/
/a/b/c/d/myTest

First list contains 5 additional elements.
First extra element 1:
/a

- ['/', '/a', '/a/b', '/a/b/c', '/a/b/c/d', '/a/b/c/d/myTest']
+ ['/a/b/c/d/myTest']
```

Maybe I'm missing something, but I had thought that using `Iterator` instead of `RawIterator` will just magically fix that issue..?

johnhaddon replied:
Maybe I didn't describe well enough (or maybe I'm wrong - let's find out!). You're correct that you only want to visit the actual match locations in the first set, the one you're iterating over. The problem I think we have is with the way we're looking up into the second set with `find()` - that is going to return a non-end iterator even if the location exists in the second set only as an ancestor. If I'm right, then including "/a/b/c" in `m1` in your unit test will give us a failure, because although "/a/b/c" doesn't exist in `m2` in its own right, it does exist there as an ancestor.

Either way, the optimisation I suggested (where we call match then `it.prune()` if there are no descendant matches) is well worth having...

mattigruener replied:
Thanks for the additional explanation :) I have pushed a commit that includes both the (previously failing) test and the new code that calls `exactMatch()`. I think that commit should be better now?

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -769,6 +773,29 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+					int i = 0;
+					for ( IECore::StringVectorData::ValueType::const_iterator it = m_linkedLights->readable().begin(); it != m_linkedLights->readable().end(); ++it, ++i )
+					{
+						// TODO: Can we avoid prefixing the path here?
```
I don't know how you feel about that kind of stuff, but I had thought we could create some kind of utility method that does the prefixing, which we would then call from both https://github.com/GafferHQ/gaffer/blob/master/src/GafferArnold/IECoreArnoldPreview/Renderer.cpp#L1362 and from here. I understand the need for prefixing - I just thought it's a little ugly to duplicate the (admittedly small) logic for prefixing? If we ever need to change the prefix, we'd have to do it twice. Maybe I'm overly concerned about something that will never come up :P

johnhaddon replied:
> I understand the need for prefixing - I just thought it's a little ugly to duplicate the (admittedly small) logic for prefixing? I don't know how you feel about that kind of stuff, but I had thought we could create some kind of utility method that does the prefixing.

Oh, I see. Personally I think I'd probably just leave it...

mattigruener commented on src/GafferScene/Preview/RendererAlgo.cpp:
```diff
@@ -512,6 +513,21 @@ struct LocationOutput
 			}
 
+			// TODO: clean up
+			CompoundObject::ObjectMap::const_iterator it = updatedAttributes->members().find( "linkedLightsSet" ); // TODO: naming needs to be determined
```
I have fixed it in one of the latest commits. I guess my reasoning was that we are pretty much done with those `attributes` after we copied over everything in https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/Preview/RendererAlgo.cpp#L507. We could just operate on the new data after that, I guess, but the latest code will use `attributes` now :)

johnhaddon replied:
But `updatedAttributes` is initialised from `m_attributes`, which is inherited from the parent. So if you do the `find()` in `updatedAttributes` you would be redoing the work for every single descendant of the location that actually had the attribute assigned.

mattigruener replied:
Ah, makes sense. Thanks! :)

mattigruener commented:
> Do you think Gaffer would ever see a "lightsToBeLinked" attribute directly? 

Are we planning on extending the lighting preview in the viewport in some way? Fancier OpenGL stuff at some point? Would we want that to respect light linking? Ha, answering your questions with new ones is probably bad style, but I guess that would be a case in which Gaffer would have to be aware of how lights and geometries interact? Maybe that would be handled in a renderer backend the same way we do it for Arnold, though...
The only other reason, that I can think of, for evaluating the expression earlier and adding the "lightsToBeLinked" attribute in the sceneGraph would be if we fail to design a good UI that operates on the SetExpressions? Passing around an explicit list of lights would probably simplify how we build the UI for light linking?

Does that make sense?

mattigruener commented on src/GafferScene/Preview/RendererAlgo.cpp:
```diff
@@ -512,6 +513,21 @@ struct LocationOutput
+
+				if( StringDataPtr lightsExpressionData = IECore::runTimeCast<StringData>( it->second ) )
+				{
+					GafferScene::ConstPathMatcherDataPtr linkedlightsSet = SetExpression::evaluateSetExpression( lightsExpressionData->readable(), scene );
```
I have added a commit that takes a first stab at the caching. I'm not sure if that's what you had in mind and I'm particularly unsure of how much you dislike my use of `mutable` in this case.

mattigruener commented:
I hope you can see all the comments I made during the day - github shows all the respective parent comments as outdated, but I guess you received an eMail about them...

johnhaddon commented:
> I hope you can see all the comments I made during the day - github shows all the respective parent comments as outdated, but I guess you received an eMail about them...

I can, by popping open the "Show outdated" sections. I've replied to a bunch of them as well, so hopefully you can see 'em by popping in there too...

> How do you feel about adding a ScenePlug::evaluateSetExpression() as an addition to ScenePlug::set() and ScenePlug::setNames()? A ScenePlug is necessary anyway for the SetExpression to make any sense.

The way I see it, set expressions are something that operate using a ScenePlug, but are not an intrinsic part of the ScenePlug itself. There are all sorts of operations that take a ScenePlug, but if they were all implemented as members then we'd end up with a very cluttered API. I've actually considered going the other way, and moving all the helper methods from ScenePlug into SceneAlgo.

I found [this article](http://www.gotw.ca/gotw/084.htm) fairly informative when thinking about this stuff. Following the "one class, one responsibility" rule I would say that the responsibility of ScenePlug is to represent a scene graph in Gaffer's plugs/contexts framework, and therefore the existing members (plug accessors, and helpers to compute those plugs in a specific context) make sense. But any higher level operations belong elsewhere.

> I was talking to Dan about it and we both felt that doing it that way could be useful as a python binding?  The test for the second stage (evaluating the expression) could be done very easily that way.

Absolutely we need it as a python binding, and we need tests. But it can just be bound as a non-member function, like the rest of SceneAlgo.

> I'll comment on the rest of your feedback as well - just wanted to get this out there while you're still around maybe :)

Sorry - I wasn't around. I wasn't officially working yesterday, but just found a bit of time to do a first pass review because I didn't want you to twiddle your thumbs...

> Are we planning on extending the lighting preview in the viewport in some way? Fancier OpenGL stuff at some point? Would we want that to respect light linking? Ha, answering your questions with new ones is probably bad style, but I guess that would be a case in which Gaffer would have to be aware of how lights and geometries interact? Maybe that would be handled in a renderer backend the same way we do it for Arnold, though...

Good question - that's another thing to inform our decision. I'm still hoping to make IPR so good that noone cares about GL, but I think it'd be reasonable for us to design something such that OpenGL _could_ respect light links if we wanted. If we did that, then I think what we'd do is reimplement the OpenGL translation with an OpenGLRenderer class as you suggested. That's actually how it was in the first iteration of Gaffer, but the original Cortex renderer API wasn't up to the job when it came to describing edits so the viewer ended up with custom code. The new API we use with Arnold is probably good enough though, so we could implement an OpenGLRenderer subclass using that, and then drive the GL viewer with the same mechanism used by the GafferScene::Preview::InteractiveRender node.

Long story short, I don't think we need to worry about GL now, because the work we do to support light linking in the InteractiveRender (I presume you're planning on adding this?) will be applicable to GL in the future.

> The only other reason, that I can think of, for evaluating the expression earlier and adding the "lightsToBeLinked" attribute in the sceneGraph would be if we fail to design a good UI that operates on the SetExpressions? Passing around an explicit list of lights would probably simplify how we build the UI for light linking? Does that make sense?

Let's call the two potential attribute forms the "expression" and the "list" for ease of explanation. I don't think there's any reason to prefer the list over the expression in Gaffer itself - after all, we can always get one from the other. The thing I'm trying to get at is that the expression attribute will inherently be a gaffer feature, requiring Gaffer's output logic to be used to get the intended render. Whereas the Renderer API is theoretically intended to be used by anyone who wants to talk to a renderer (a procedural for instance, or someone writing a Gaffer alternative). In that scenario only the list attribute makes sense, which means there may be workflows where cache files store the list attribute. Which means that when loaded in Gaffer we might see the list attribute alongside the expression attribute, with all the attendant ambiguities that brings.

Actually, I've just realised that if we add the "/name/a/light/directly/in/place/of/a/set" syntax I proposed to the expression language, then the expression is a just a superset of the list.

I think I'm leaning towards making the list and the expression attribute be the very same thing, with the same name ("linkedLights" or "lightLinks"?) in the Renderer API and StandardAttributes node. So it can be used naively elsewhere, but has added bells and whistles when used in Gaffer.

This brings me to another thought. It's going to be a pain to implement the light link outputs for the InteractiveRender node as well as the Render node, and the caching in the RenderSets version needs work anyway. So maybe it makes sense to instead implement an EvaluateLightLinks node that is used internally by both? And then it could also be used explicitly if we ever wanted to generate a .scc or .abc file with naive lists of lights so they could be used outside Gaffer?

johnhaddon commented on src/GafferScene/Preview/RendererAlgo.cpp:
```diff
@@ -513,18 +532,20 @@ struct LocationOutput
-					GafferScene::PathMatcher linkedlightsSet = SetExpression::evaluateSetExpression( lightsExpressionData->readable(), scene );
-					linkedlightsSet.paths( lightNamesWritable );
-					updatedAttributes->members()["lightsToBeLinked"] = lightNames; // TODO: naming needs to be determined
+					IECore::StringVectorDataPtr lightNames = new IECore::StringVectorData();
```
I should have mentioned that we should probably use the `IECore::LRUCache` to simplify the caching. This would have taken care of the two big problems in this implementation :

1. Multiple threads are reading/writing m_lightLinkExpressions at the same time, which will lead to crashes.
2. Multiple threads may evaluate the same expression because it wasn't in m_lightLinkExpressions when they checked, but was by the time they finished their own evaluation.

Anyway, I've suggested in my main comment that perhaps we should have an EvaluateLightLinks node so we don't need to do anything in the render output layer - let's resolve that discussion before worrying more about the caching.

johnhaddon commented:
> This brings me to another thought. It's going to be a pain to implement the light link outputs for the InteractiveRender node as well as the Render node, and the caching in the RenderSets version needs work anyway. So maybe it makes sense to instead implement an EvaluateLightLinks node that is used internally by both?

On this topic, here's a branch from a while back where I was prototyping a mechanism for registering such adaptors :

https://github.com/johnhaddon/gaffer/tree/renderAdaptors

mattigruener commented:
Thanks for linking to that article. Good read!

> Sorry - I wasn't around. I wasn't officially working yesterday, but just found a bit of time to do a first pass review because I didn't want you to twiddle your thumbs...

Thanks for replying anyway! I have other stuff on my list in case you don't have time to reply :)

> Do you intend to add any more features to the expression language to this PR? I was thinking parenthesis for grouping might be nice, and possibly also treating a "space separated list of sets" the same way as a "pipe | separated | list | of | sets" to provide simple functionality for folks with expression blindness. 

I am happy to add more to it, but didn't really intend to for this PR. The space separated list is a great idea, though and I'll look into it. Considering the rest of the discussion about using just a single attribute for both list and expression, it sounds like we will need it for this PR anyway?

Can you give me an example regarding the parenthesis for grouping? Something like this

`( (A | B) & C ) | D`

should already work, I think. Is that what you mean?

> Perhaps also an "/explicitly/named/light" syntax might be handy?

> Actually, I've just realised that if we add the "/name/a/light/directly/in/place/of/a/set" syntax I proposed to the expression language, then the expression is a just a superset of the list.

Ok, I have a question about that. I just tried and "/name/a/light/directly/in/place/of/a/set" is a perfectly valid name for a set in Gaffer. I guess what we could do in order to distinguish between sets and lights is to interpret a missing set [here](https://github.com/mattigruener/gaffer/blob/2e2a91770a5a4f76ab1a01eac514d09b555535a8/src/GafferScene/SetExpression.cpp#L148) as a potentially valid light location and create a `PathMatcher` for it? Maybe only if it starts with a "/"? You didn't really like printing stuff from there anyway - I still feel like it could be confusing to the user, especially if that's how we solve the set vs. light problem. What was your idea for differentiating between a light location and a (admittedly obscure) set name that looks like one? Am I missing something?

> I think I'm leaning towards making the list and the expression attribute be the very same thing, with the same name ("linkedLights" or "lightLinks"?) in the Renderer API and StandardAttributes node. So it can be used naively elsewhere, but has added bells and whistles when used in Gaffer.

I like "linkedLights" for a list of lights better, but I don't feel strongly about it either way. 

> So maybe it makes sense to instead implement an EvaluateLightLinks node that is used internally by both? And then it could also be used explicitly if we ever wanted to generate a .scc or .abc file with naive lists of lights so they could be used outside Gaffer?

Using just one attribute with "added bells and whistles" in Gaffer sounds good to me. I will look into adding the EvaluateLightLinks node and moving the expression eval stuff into it from the renderer back end. Am I right in assuming that we could potentially run in a situation where some .scc brings in a list of lights and we then unnecessarily apply all the bells and whistles only to parse/eval that list into the same list again? Any way to determine that we don't have to (except just doing it once and then rely on hashing)?

Thinking about .scc files bringing in a list that contains all kinds of goofy lights that probably don't even exist in the scene, I realize that I should definitely check for non-existing lights [here](https://github.com/mattigruener/gaffer/blob/2e2a91770a5a4f76ab1a01eac514d09b555535a8/src/GafferArnold/IECoreArnoldPreview/Renderer.cpp#L790) in order to not clutter that attribute with non-existing nodes. In general, do you feel like it's a useful workflow to store the links in that cache? I'm just wondering if that could lead to weird situations where someone innocently renames a light in his scene and accidentally hits a link that was brought in by a cache only to mess up their lighting completely...

Will also look into your other suggestions and add respective commits.

Thanks :)

mattigruener commented:
Added a few updates from today.

mattigruener commented:
> Do you think we can just disallow a / in the name in the Set node, or do you think we have examples of these / prefixed sets in the wild?

I haven't seen any.

> I was just assuming we'd embed it straight into the set syntax that something beginning with / was always an explicit path, and never a set name. I think that's better than allowing ambiguity in the parser and then having to resolve it in the visitor. 

Agreed. I have added support for objectNames now, but the visitor still has to do some of the work in the current implementation. Both setNames and objectNames are just stored as strings at the parser stage... I started reworking some of that towards the end of today to use structs for SetName and ObjectName - will continue working on that tmrw.

> I don't think we need it in this PR, but it would be nice to have if you get the chance.

I've added a first version of it while I looked at support for objectNames.

> Could you add a test demonstrating it in action?

There's a bunch of new tests now that show grouping and the new features. I kinda like that we can effectively define unnamed sets now by doing something like:

`A & (/group/group/sphere1 /group/group/sphere42)`

But I guess that also has the potential to get wild and messy fairly quickly? How do you feel about the current set of features of our little language? Now that we have support for objectNames, it feels like one necessary next step would be adding support for something like this:

`bagOfObjects - objectWeDontWant`

Dan and me also talked about supporting wildcards for the object names, so that `/car/headlights/*` actually expands to `(/car/headlights/leftLight /car/headlights/rightLight)` during evaluation. That could be quite expensive, though. To offset some of that cost, the function that does the evaluation could take a set of objects that define the ones we are interested in at all (all lights, for example)? We would then loop over all those and check if they match the wildcard. Maybe an addition for later? Seemed quite useful to us to support wildcards...

> Actually I realise I don't know the operator precedence of | and & here - is it implemented so that & takes precedence over | as in C++?

Currently it's not even accepting something like

`A | B & C`

but needs the explicit grouping via (). I realize it's not ideal and I will look into lifting that restriction. Will make things a little more complicated, I guess, but requiring the parenthesis makes the whole thing very lisp'y, doesn't it? ;)

> I guess this is a fairly straightforward lightSet.intersection( resultOfMyExpression ) though?

Yes, true. General question: Sets are supposed to contain every match explicitly and we don't consider children of objects in the set as part of the set? I'm just asking because I'm wondering if adding `/car/headlights` to a set should make the children come through as linked lights.

> How do you want to progress with the renderAdaptors thing?

Good question - I think it would make sense to get that cleaned up anyway? It's useful not just for this application, but we could use it to remove a lot of the translator stuff that we're currently doing at IE, I guess? I am happy to look into it - what did you expect what we'd still have to do on it? Even if we drop the Evaluation node (whatever we might call it) into all IERenderPasses - I find it very odd that standalone users will have to remember to add that node to their graphs to get their expressions evaluated? Feels like something that should be taken care of behind the scenes...

I didn't get to adding a test that checks the linked lights on the Arnold side, but I'll look into that tomorrow. I hope I didn't forget about any of today's open questions...

Thanks!

mattigruener commented:
What happened to that latest comment? :P

I didn't even see it this morning, because I am the author of it? Just noticed when I pushed the change that preserves knowledge about setNames and objectNames...

mattigruener commented:
>> bagOfObjects - objectWeDontWant

> That was the one obvious omission I could see too. It seems like an easy one to add, since you already have your BinaryOp class and we already have PathMatcher.removePaths( otherPathMatcher ), so could you get it in for this PR do you think?

Yes, will do. Should be fairly straight forward although it will need some reshuffling - especially keeping in mind that I have to work on the operator precedence as well.

> Does that mean that A | B | C is also a parse error?

Yeah, none of those will currently work. Will look into it. I just added a test that should uncover this problem.

> This raises the question of whether or not we're going to support light linking for the 3delight renders too? Personally I feel like we'd be better off saving our effort and hoping to save up some time to add a shiny new NSI or PRMan21 backend to replace the legacy ones. In that case, I could just remove the WIP commit and put the renderAdaptors branch up for review.

I think that would be best. If it should ever come up we can still add it for those nodes, but I'd personally leave it until that happens... I haven't looked at the Appleseed implementation at all, but supporting that would make more sense to me.

> Great! Seems like we're getting close?

I think so! With some of the changes that I still need to make I don't know how much more boost::spirit pain they will cause and how long it'll take to sort that out. But that went somewhat smooth so far...

I guess we'll handle a UI for lightlinking in a different PR?

johnhaddon commented:
>What happened to that latest comment? :P

Sorry! Looks like I did my usual muppet thing and opened your comment for editing so I could get the original markdown to cut + paste, and then forgot the cut + paste step. I've tried to find a way of disabling edits but there doesn't seem to be one...

> I guess we'll handle a UI for lightlinking in a different PR?

Yes, but I would like to add the linkedLights (lightLinks? maybe a vote?) attribute to the StandardAttributes node, so we have basic support for setting them.

I'll get my renderAdaptors branch up for review today. Since we seem to have a good plan in place now, I think I'll wait until you have all your ducks in a row and your commits squashed before I do a last review (unless you want specific feedback on anything in the meantime). Sounds OK?

mattigruener commented:
Yes, sounds good :) I'll let you know once I feel like I am done. Thanks!

mattigruener commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -0,0 +1,455 @@
+	qi::rule<Iterator, ExpressionAst(), ascii::space_type> nameOrListOfNames, listOfNames, objectNameOrSetName, element;
+};
+
+void printSyntaxError( const std::string &setExpression, std::string::const_iterator position )
```
Not sure if you like the way I handle the errors here, @johnhaddon?

```
WARNING : GafferScene::SetAlgo : Syntax error in indicated part of SetExpression.
setA - (/group/group/sphere2
     |---------------------|
```

Adding support for better errors would be a little tricky I guess. We could potentially run a check for unbalanced parantheses here, but meh.

mattigruener replied:
I'm using this example again in the following:

```
WARNING : GafferScene::SetAlgo : Syntax error in indicated part of SetExpression.
setA - (/group/group/sphere2
     |---------------------|
```

So yeah, I thought about this one for a bit when I implemented it and in theory I prefer the ^ way of indicating an error. But in this case it's not really saying "Hey, that minus looks off!", but it's saying "Ok, I know that setA is a set name and you're allowed to use a minus here, but only if your second operand matches what I know are valid operands for andNot operations. What you're doing there is weird and it doesn't match what I would expect a valid operand. Something in what you're doing following the set name is fishy". 

That's why I felt that simpy pointing to the minus is misleading, because really it's all the rest that spirit has a problem with. Does that make sense?

mattigruener replied:
Maybe a less wordy way of saying this is that spirit indicates up to which point parsing according to the rules in the grammar succeeded. Showing that to the user in the way that clang does it might be misleading?

johnhaddon replied:
OK, that makes sense - thanks for explaining. Let's go with what you have then but throw an exception rather than post a message...

mattigruener commented on include/GafferScene/SetAlgo.h:
```diff
@@ -0,0 +1,62 @@
+PathMatcher evaluateSetExpression( const std::string &setExpression, const ScenePlug* scene );
+
+IECore::MurmurHash setExpressionHash( const std::string &setExpression, const ScenePlug* scene );
+void hashSetExpression( const std::string &setExpression, const ScenePlug* scene, IECore::MurmurHash &h );
```
I added both of these because they both seem useful. Do I violate Gaffer conventions by doing that?

mattigruener commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -0,0 +1,455 @@
+std::ostream& operator << (std::ostream& stream, const ExpressionAst& expr)
+{
+	AstPrinter printer;
+	printer(expr);
```
This is a little messy, I guess, because it's not guaranteed that the `AstPrinter` instance writes into `stream`. We could add a constructor for `AstPrinter` that stores a given `std::ostream` reference, but I'm not sure it's worth it for simple debugging stuff.

mattigruener commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -0,0 +1,455 @@
+		BOOST_SPIRIT_DEBUG_NODE(objectName);
+		BOOST_SPIRIT_DEBUG_NODE(nameOrListOfNames);
+		BOOST_SPIRIT_DEBUG_NODE(listOfNames);
+		BOOST_SPIRIT_DEBUG_NODE(objectNameOrSetName);
```
Do you want to leave this in here? I commented out the `#define` at the top of the file, so these don't do anything. I found this quite useful for debugging, but I'm not sure what to expect regarding how often we will have to touch this.

mattigruener commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -0,0 +1,455 @@
+		AstPrinter printer;
+		printer(ast);
+		std::cout << "\n-------------------------\n";
+		#endif
```
Same as above - we can get rid of this, but I found it useful for debugging.

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -487,6 +487,8 @@ IECore::InternedString g_dispAutoBumpAttributeName( "ai:disp_autobump" );
 IECore::InternedString g_curvesMinPixelWidthAttributeName( "ai:curves:min_pixel_width" );
 IECore::InternedString g_curvesModeAttributeName( "ai:curves:mode" );
 
+IECore::InternedString g_linkedLights( "gaffer:linkedLights" );
```
Do you want to store this somewhere more global? We use this in different places...

mattigruener commented on src/GafferArnold/IECoreArnoldPreview/Renderer.cpp:
```diff
@@ -771,6 +775,34 @@ class ArnoldAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
+					for( std::vector<AtNode*>::size_type i = 0; i < lightNodesVector.size(); ++i )
+					{
+						AiArraySetPtr( linkedLightNodes, i, lightNodesVector[i] );
+					}
```
Avoiding NULL pointers here might be unnecessary because we intersect with the light set in `EvaluateLightLinks` now. Do you still want to keep this?

mattigruener commented on python/GafferSceneTest/EvaluateLightLinksTest.py:
```diff
@@ -0,0 +1,119 @@
+			evalNode["out"].attributes( "/group/sphere" )[ "gaffer:linkedLights" ],
+			IECore.StringVectorData( [ "/group/group/light", "/group/group/light1" ] ) )
+
+		# Make sure changing sets updates the result of the expression evaluation
```
I was wondering if you'd like these to be separate test methods? Do we have a convention for setting up a test script and using it in mulitple tests? Creating the entire graph over and over again for every test seems a little ugly.

johnhaddon commented on src/GafferScene/EvaluateLightLinks.cpp:
```diff
@@ -0,0 +1,127 @@
+{
+	SceneProcessor::hashAttributes( path, context, parent, h );
+
+	inPlug()->attributesPlug()->hash( h );
```
I think we can be friendlier to Gaffer's cache here - by passing through an unchanged hash when the light links attribute doesn't exist we can allow Gaffer to share cache entries with our input for anything we don't touch. Something like this (untested!)...

```
const MurmurHash inputHash = inPlug()->attributesPlug()->hash();
ConstCompoundObjectPtr attributes = inPlug()->attributesPlug()->getValue( & inputHash );
ConstStringDataPtr expressionData = attributes->member<StringData>( "linkedLights" );
if( !expressionData )
{
    // Pass through.
    h = inputHash;
    return;
}

SceneProcessor::hashAttributes( path, context, parent, h );
h.append( inputHash );
h.append( SetAlgo::setExpressionHash( expressionData->readable(), inPlug() );
```

johnhaddon commented on src/GafferScene/EvaluateLightLinks.cpp:
```diff
@@ -0,0 +1,127 @@
+			SetAlgo::hashSetExpression( lightsExpressionData->readable(), inPlug(), h );
+		}
+	}
+  // IECore::ConstInternedStringVectorDataPtr setNames = inPlug()->setNamesPlug()->getValue();
```
Remove me!

johnhaddon commented on src/GafferScene/EvaluateLightLinks.cpp:
```diff
@@ -0,0 +1,127 @@
+	IECore::ConstCompoundObjectPtr attributes = inPlug()->attributesPlug()->getValue();
+
+	CompoundObject::ObjectMap::const_iterator it = attributes->members().find( m_lightLinkAttrName );
+	if( it != attributes->members().end() )
```
I'm a big fan of returning early from functions rather than nesting deeper and deeper into conditionals like this - it makes the code much easier to follow, because you're not having to maintain a mental stack of "reasons I'm in this code path". Mirroring the hash example code above would also make it much easier to verify they match :

```
ConstCompoundObjectPtr inputAttributes = inPlug()->attributesPlug()->getValue( & inputHash );
ConstStringDataPtr expressionData = attributes->member<StringData>( "linkedLights" );
if( !expressionData )
{
    // Pass through
    return inputAttributes;
}

// Real work here, no nesting.
```

johnhaddon commented on src/GafferScene/EvaluateLightLinks.cpp:
```diff
@@ -0,0 +1,127 @@
+	{
+		if( StringDataPtr lightsExpressionData = IECore::runTimeCast<StringData>( it->second ) )
+		{
+			IECore::CompoundObjectPtr result = attributes->copy();
```
[This](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/ShaderAssignment.cpp#L107) method of copying provides enough of a performance boost that it's worth using instead.

johnhaddon commented:
Oh, forgot to say, for the registering of the adaptor, how about we just add a `startup/GafferScene/adaptors.py` file and put the registration there? Then it's also providing a reasonably discoverable customisation example for anyone who might also want to tinker with adaptors...

mattigruener commented:
Thanks for the feedback! I'll try to get to most of it today after I finished some other stuff that came up. That'll definitely be enough to keep me busy until you have time to continue the review :)

mattigruener commented on include/GafferScene/SetAlgo.h:
```diff
@@ -0,0 +1,62 @@
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFER_SETEXPRESSION_H
```
The way I think about this is as a new, separate unit of functionality, but that's probably mostly because I didn't implement the other stuff... I totally get if you want to keep the two things together. I am happy to integrate it into `SceneAlgo`. Alternatively, how about moving the set related functionality into `SetAlgo`? That would probably force us to touch all the spots where its used, which might be annoying... It's just that SceneAlgo feels quite massive already?

I have fixed that typo for now and will update the code once we make a decision here.

johnhaddon replied:
OK, let's leave it in SetAlgo then, and leave it as a todo to move the rest over from SceneAlgo.

mattigruener commented on src/GafferScene/EvaluateLightLinks.cpp:
```diff
@@ -0,0 +1,127 @@
+{
+	SceneProcessor::affects( input, outputs );
+
+	if( input == inPlug()->attributesPlug() || input == inPlug()->setPlug() || input == inPlug()->globalsPlug() )
```
I like the `SetAlgo::affectsSetExpression` idea and have added it in one of the commits I'll be pushing in a bit. Could you have a look? One thing I'm not sure of is whether it's ok to just check the `Direction` and by doing so allowing all inputs to affect the outputs as opposed to what the node did before by looking at `inPlug()` only. For `EvaluateLightLinks` it doesn't matter, but it might for a node that we come up with in the future? Without checking the `Direction` I got errors because the `attributesPlug` of the `outPlug` would then depend on itself. 
We could just pass in the parent explicitly, but I kinda liked the clean function signature. Do you see a problem with potentially depending on multiple inputs? @danieldresser pointed out that in the cases in which we have multiple inputs we might actually want to depend on all of them anyway, because a merge node for example would use all the data that is flowing in.

johnhaddon replied:
I've made a bunch of inline comments about this - basically I think you were confusing the current needs of EvaluateLightLinks with the more general needs of other clients of set expressions.

mattigruener commented:
Hey John,

finally back to light linking. I haven't had time today to address all the changes you requested, but I guess I fixed most of the problems that you found and will do the rest tomorrow. I also replied to now outdated comments - so you might have to open those explicitly again.

Thanks! :)

Matti.

johnhaddon commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -449,6 +449,22 @@ IECore::MurmurHash setExpressionHash( const std::string &setExpression, const Sc
+bool affectsSetExpression( const Plug *scenePlugChild )
+{
+	const ScenePlug *parent = scenePlugChild->parent<ScenePlug>();
+	if( !parent || parent->direction() != Gaffer::Plug::In )
```
I don't think we need the direction check here, since set expressions won't always be used only internally to a node.  For instance, a UI element might be monitoring an output plug to show the results of a set expression.

johnhaddon commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -449,6 +449,22 @@ IECore::MurmurHash setExpressionHash( const std::string &setExpression, const Sc
+		return false;
+	}
+
+	if( parent->attributesPlug() == scenePlugChild || parent->setPlug() == scenePlugChild )
```
Does the attributesPlug really affect the result of a set expression? I think this is mixing the implementation of the EvaluateLightLinks node with the more general-purpose set expression code.

johnhaddon commented on src/GafferScene/EvaluateLightLinks.cpp:
```diff
@@ -70,7 +70,7 @@ void EvaluateLightLinks::affects( const Gaffer::Plug *input, AffectedPlugsContai
 	SceneProcessor::affects( input, outputs );
 
-	if( input == inPlug()->attributesPlug() || input == inPlug()->setPlug() || input == inPlug()->globalsPlug() )
+	if( SetAlgo::affectsSetExpression( input ) )
```
We should keep the attributesPlug check here (and take it out of affectsSetExpression), and also check that `plug->parent() == inPlug()` before calling `affectsSetExpression()`.

johnhaddon commented:
> I haven't had time today to address all the changes you requested, but I guess I fixed most of the problems that you found and will do the rest tomorrow. I also replied to now outdated comments - so you might have to open those explicitly again.

Thanks! I think I've replied to all your comments, but do let me know if I've missed one...

mattigruener commented:
I just pushed a few new commits. Most are just to address your feedback, but I also restructured and improved the grammar a little because after starting to play with it in Gaffer I immediately found something that behaved weirdly. Should be better now and easier to understand too. Thanks Don, for helping with that.

I would expect tests to go through now...

mattigruener commented:
I have added a first draft for the documentation. I'll go through the examples again tomorrow with a less tired brain and make sure that they are actually correct.

johnhaddon commented:
Thanks Matti. The tests are because we haven't merged the renderAdaptors branch - I'll do that now. Looks like we've also got a compilation error with some of the compilers :

```
src/GafferScene/SetAlgo.cpp:201:15: error: unused function 'operator<<' [-Werror,-Wunused-function]
std::ostream& operator<<( std::ostream& stream, const ExpressionAst& expr )
```

Could you push a fix for that and then I'll take a final look once the tests have passed?
Cheers...
John

mattigruener commented:
So, I've been trying to get the tests here and in your other PR to pass throughout the day today and I'll just restart the currently failing job now and then call it a day. Right now, it's failing with what looks like a fairly severe crash on the appleseed side. I created a [gist](https://gist.github.com/mattigruener/14b38abca6fe04008d07fb7b36220deb) to preserve that output.

johnhaddon commented:
I mailed Esteban about the GafferAppleseed test and he thinks it's most likely a threading problem with GafferAppleseed. I'm not sure why it's showing up more now, but maybe just because the presence of an adaptor is changing the timings somewhat? He's going to take a look when he gets a chance...

I also learned that the next Appleseed will be the last one before they require C++11. I think it's time to draw a line in the sand ourselves...

mattigruener commented:
Thanks for looking into it :) C++11 would be great for sure... I guess there's not much left for me to do on this PR for now - do you want me to start cleaning up the history or would you like to keep it for the review?

johnhaddon commented:
> do you want me to start cleaning up the history or would you like to keep it for the review?

Squash it please, let's get this one out there!

mattigruener commented:
I think the history should be pretty clean now.

johnhaddon commented:
Nice one! Thanks Matti, and thanks @donboie for the (boost) spiritual guidance!

mattigruener commented:
Woohoo, thanks guys :)
